[GSoC] Coroutines migration PreloadNextCard

divyansh-dxn opened the pull request:
## Fixes
A part of #7108

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

divyansh-dxn commented:
Squash Merge. Made separate commits just for easy review.

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:
```diff
@@ -410,8 +408,12 @@ abstract class AbstractFlashcardViewer :
-        // listener and not in background
+        launchCatchingTask {
+            withCol {
+                sched.count() // Ensure counts are recomputed if necessary, to know queue to look for
```
This doesn't replicate the original code.

In the original code if there was an error we would log it to logcat and just ignore it as the reviewing can continue.
In your code if an error would happen a dialog with the exception would appear. This dialog isn't needed because reviewing can continue, it breaks the user flow(with the dialog showing) and the user can't do anything about the error(and it would be confusing).

So I would ask you to keep the original code here and wrap the two lines in the try-catch.

Arthur-Milchior replied:
By itself, I’m fine with such a change. 
I don’t think we are actually expecting any exception, I’m not even convinced we actually had exceptions here, or that we can have exceptions with preloading a card that won’t anyway be an exception with loading a card. After all, the only exceptions I can imagine are when database become inaccessible.
This GSoC project is a great occasion to do such cleaning.

However, I agree with @lukstbit that a change is due. Doing the whole migration while keeping the try catch. Then, in a last commit, removing it and explaining why it is probably useless. This way, if we are wrong, we can easily revert this last commit and keep the migration.

Of course it means it would not be a squash-merge anymore. However, I’m fine with the fact that you move and delete in the same commit.

divyansh-dxn commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:
```diff
@@ -1150,7 +1150,8 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
         override fun task(col: Collection, collectionTask: ProgressSenderAndCancelListener<Void>): Void? {
             try {
-                col.sched.counts() // Ensure counts are recomputed if necessary, to know queue to look for
+                // TODO: This function has been migrated to coroutines and no longer used, so remove this class
```
Yeah, sorry I missed. It was supposed to be above class.

divyansh-dxn commented:
Done the migration and deleted the unused class in same commit. And removed the try-catch in a saparate commit with a note, so in future if something arises due to this, can be reverted.
